### PR TITLE
[WIP] [Train] Pytorch microbenchmark

### PR DESCRIPTION
**Pull request by amogkam**
Adds microbenchmark comparing Ray Train performance vs. `torch.distributed`

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: release/train_tests/microbenchmarks/workloads/torch_microbenchmark.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+    total_time = time.time() - current_time
+    print("Total time for Ray Train: ", total_time)
+    return total_time
+
+
+def single_node_benchmark():
+    import ray
+    ray.init("auto")
+
+    assert NUM_NODES * NUM_GPUS_PER_NODE == NUM_GPUS_PER_NODE
```
Comment (matthewdeng) [inline release/train_tests/microbenchmarks/workloads/torch_microbenchmark.py:156]:
```suggestion
    assert NUM_NODES == 1
```

File: release/train_tests/microbenchmarks/workloads/torch_microbenchmark.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+    port = find_free_port()
+    torch_time = torch_run_single_node(
+        train_func=torch_train,
+        world_size=NUM_NODES * NUM_GPUS_PER_NODE,
+        num_workers_per_node=NUM_GPUS_PER_NODE,
+        address=address,
+        port=port)
+
+    # Time using Ray Train.
+    ray_time = ray_train_run(train_func=train_func, world_size=4)
```
Comment (matthewdeng) [inline release/train_tests/microbenchmarks/workloads/torch_microbenchmark.py:169]:
```suggestion
    ray_time = ray_train_run(train_func=train_func, world_size=NUM_NODES * NUM_GPUS_PER_NODE)
```

**Review by matthewdeng — COMMENTED** [review]
Only took a cursory glance so far, but so far this is really cool!

I think it would be worthwhile for us to consider upstreaming this to a tutorial in the documentation later, which might even go a step further and use `torch.distributed.run` instead of Ray tasks to execute the script on each node. It feels almost magical to be able to run distributed code through a single python script 😄

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Pull request closed without merging**